packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,381 @@
+  connectionManager: Libp2p["connectionManager"];
+  peerStore: Libp2p["peerStore"];
+  ping: Libp2p["ping"];
+  dial: Libp2p["dial"];
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+export interface Options {
+  /**
+   * Number of attempts before a peer is considered non-dialable
+   * This is used to not spam a peer with dial attempts when it is not dialable
+   */
+  maxDialAttemptsForPeer?: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   */
+  maxBootstrapPeersAllowed?: number;
+  // TODO: add this functionality back
+  // /**
+  //  * Number of attempts before dialling bootstrap nodes over the {@link maxBootstrapPeersAllowed} value.
+  //  * Dialing is the process of opening an outgoing connection to a listening peer with a transport that is supported by both peers.
+  //  * This is only used when other discovery mechanisms haven't yield peers that are dialable
+  //  * This increases relative centralization of the network
+  //  */
+  // maxTimeBeforeBootstrapFallbackMs?: number;
+  /**
+   * Interval at which to log the state of the connection manager
+   */
+  loggingIntervalMs?: number;
+}
+
+export class ConnectionManager {
+  private static instances = new Map<string, ConnectionManager>();
+  private keepAliveManager = new KeepAliveManager();
+  private options: Options;
+  private libp2pComponents: Libp2pComponents;
+  private dialAttemptsForPeer: Map<string, number> = new Map();
+
+  public static create(
+    peerId: string,
+    libp2p: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ): ConnectionManager {
+    let instance = ConnectionManager.instances.get(peerId);
+    if (!instance)
+      instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+    ConnectionManager.instances.set(peerId, instance);
+
+    return instance;
+  }
+
+  constructor(
+    libp2pComponents: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ) {
+    this.libp2pComponents = libp2pComponents;
+    this.options = options ?? {};
+
+    this.runService(keepAliveOptions, relay)
+      .then(() => log(`Connection Manager is now running`))
+      .catch((error) => log(`Unexpected error while running service`, error));
+  }
+
+  private async runService(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): Promise<void> {
+    const { loggingIntervalMs = DEFAULT_LOGGING_INTERVAL_MS } = this.options;
```

Comment:
Since this function is essentially initializator of the class it would be better to hide this function behind `if initialized` condition if it is not intended to be used otherwise.

Reply:
That much I see :) What I want to prevent here is further possible misunderstanding of the method `runService` and it's accidental usage. Also, it's not bulletproof to rely on TS since our consumers might use plain JS and nothing will prevent them from doing `instance.runScript()`

Reply:
Of course we don't want to go into this rabbit hole. If you think it is not needed to handle this edge case here - feel free to resolve this point.

I think `if instance` is good enough. But because `runScript` is separated into a method it usually implies that it can or will be used in some other cases apart from the original one in constructor.

Reply:
I have commented similarly on https://github.com/waku-org/js-waku/pull/1145#discussion_r1098115537

Applying defaults every time the `options` is read is not ideal because it is prone to error.

I would also recommend to set the default in the `constructor`. If that's not convenient (only reason I can think off is because we want to preserve what was originally passed but does not seem to be the case here) then provide a method that applies default values and call this method anytime `options` needs to be read.